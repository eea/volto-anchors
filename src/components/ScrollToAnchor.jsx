import React from 'react';
import { connect } from 'react-redux';
import {
  openAccordionIfContainsAnchors,
  waitForElm,
  scrollToTarget,
} from '@eeacms/volto-anchors/helpers';

const ScrollToAnchor = (props) => {
  const { location } = props;

  React.useEffect(() => {
    const anchor = location?.hash || location?.pathname.hash;
    if (anchor) {
      waitForElm(anchor)
        .then((elm) => {
          scrollToTarget(elm);
        })
        .catch((err) => {
          // eslint-disable-next-line no-console
          console.error(err);
        });
      openAccordionIfContainsAnchors(anchor);
    }
  }, [location]);

  return null;
};

export default connect(
  (state, props) => ({
    location: state.loadProtector.location,
  }),
  {},
)(ScrollToAnchor);
