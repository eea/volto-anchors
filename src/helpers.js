import { isArray } from 'lodash';
import Slugger from 'github-slugger';
import config from '@plone/volto/registry';
import { serializeNodes } from '@plone/volto-slate/editor/render';
import { Link } from 'react-router-dom';
import { getBlocks } from '@plone/volto/helpers/Blocks/Blocks';
import linkSVG from '@plone/volto/icons/link.svg';

import './less/slate-anchors.less';

export const createSlateParagraph = (text) => {
  return isArray(text) ? text : config.settings.slate.defaultValue();
};

export const serializeText = (text) => {
  return isArray(text) ? serializeNodes(text) : text;
};

export const toSlug = (url) => Slugger.slug(url);

export const waitForElm = (selector) => {
  return new Promise((resolve) => {
    if (document.querySelector(selector)) {
      return resolve(document.querySelector(selector));
    }

    const observer = new MutationObserver((mutations) => {
      if (document.querySelector(selector)) {
        resolve(document.querySelector(selector));
        observer.disconnect();
      }
    });

    observer.observe(document.body, {
      childList: true,
      subtree: true,
    });
  });
};

export const scrollToTarget = (target, offsetHeight = 0) => {
  const bodyRect = document.body.getBoundingClientRect().top;
  const targetRect = target.getBoundingClientRect().top;
  const targetPosition = targetRect - bodyRect - offsetHeight;

  window.scrollTo({
    top: targetPosition,
    behavior: 'smooth',
  });

  return;
};

export const openAccordionIfContainsAnchors = (anchor) => {
  waitForElm(anchor).then((elm) => {
    if (elm.closest('.accordion')) {
      const comp = elm.closest('.accordion')?.querySelector('.title');
      if (!comp?.className?.includes('active')) {
        comp.click();
        setTimeout(() => scrollToTarget(elm), 300);
      }
    }
  });

  return;
};

//post order traversal of blocks content

export const visitBlocks = (content, callback) => {
  const stack = getBlocks(content);
  while (stack.length > 0) {
    const [id, blockdata] = stack.pop();
    const wantBreak = callback([id, blockdata]);
    if (wantBreak) break;
    // assumes that a block value is like: {blocks, blocks_layout} or
    // { data: {blocks, blocks_layout}}
    if (Object.keys(blockdata || {}).indexOf('blocks') > -1) {
      stack.push(...getBlocks(blockdata));
    }
    if (Object.keys(blockdata?.data || {}).indexOf('blocks') > -1) {
      stack.push(...getBlocks(blockdata.data));
    }
  }
};

export const renderLinkElement = (tagName) => {
  function LinkElement({ attributes, children, mode = 'edit', className }) {
    const Tag = tagName;
    const slug = attributes.id || '';
    const { slate = {} } = config.settings;

    return slate.useLinkedHeadings === false ? (
      <Tag className={className} {...attributes}>
        {children}
      </Tag>
    ) : (
      <Tag className={className} {...attributes}>
        {mode === 'view' && slug && (
          <Link
            className="anchor"
            aria-hidden="true"
            tabIndex={-1}
            to={`#${slug}`}
          >
            <svg
              {...linkSVG.attributes}
              dangerouslySetInnerHTML={{ __html: linkSVG.content }}
              width="2em"
              height={null}
            ></svg>
          </Link>
        )}
        {children}
      </Tag>
    );
  }
  LinkElement.displayName = `${tagName}LinkElement`;
  return LinkElement;
};
