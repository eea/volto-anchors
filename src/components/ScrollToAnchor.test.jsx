import React from 'react';
import { Provider } from 'react-intl-redux';
import configureStore from 'redux-mock-store';
import { render } from '@testing-library/react';
import ScrollToAnchor from './ScrollToAnchor';
import { openAccordionIfContainsAnchors } from '@eeacms/volto-anchors/helpers';

const mockStore = configureStore();

jest.mock('../helpers', () => ({
  openAccordionIfContainsAnchors: jest.fn(),
}));

describe('ScrollToAnchor', () => {
  const store = mockStore({
    intl: {
      locale: 'en',
      messages: {},
    },
    loadProtector: {
      location: {
        hash: '#anchor',
        pathname: {
          hash: '#anchor',
        },
      },
    },
  });

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should call openAccordionIfContainsAnchors with the correct hash', () => {
    render(
      <Provider store={store}>
        <ScrollToAnchor />
      </Provider>,
    );
    expect(openAccordionIfContainsAnchors).toHaveBeenCalledWith('#anchor');
  });

  it('should call openAccordionIfContainsAnchors with the pathname hash', () => {
    const newStore = mockStore({
      intl: {
        locale: 'en',
        messages: {},
      },
      loadProtector: {
        location: {
          pathname: {
            hash: '#anchor',
          },
        },
      },
    });
    render(
      <Provider store={newStore}>
        <ScrollToAnchor />
      </Provider>,
    );
    expect(openAccordionIfContainsAnchors).toHaveBeenCalledWith('#anchor');
  });

  it('should not call openAccordionIfContainsAnchors', () => {
    const newStore = mockStore({
      intl: {
        locale: 'en',
        messages: {},
      },
      loadProtector: {
        location: {
          pathname: {
            hash: null,
          },
        },
      },
    });

    render(
      <Provider store={newStore}>
        <ScrollToAnchor />
      </Provider>,
    );
    expect(openAccordionIfContainsAnchors).not.toHaveBeenCalled();
  });
});
