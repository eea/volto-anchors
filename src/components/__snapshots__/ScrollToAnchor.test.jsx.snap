// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScrollToAnchor should match the snapshot 1`] = `null`;
